torch/distributions/utils.py

```diff
@@ -1,9 +1,31 @@
+from collections import namedtuple
 from functools import update_wrapper
 from numbers import Number
 
 import torch
-from torch.autograd import Variable
 import torch.nn.functional as F
+from torch.autograd import Variable
+
+# This follows semantics of numpy.finfo.
+_Finfo = namedtuple('_Finfo', ['eps', 'tiny'])
+_FINFO = {
+    'torch.HalfTensor': _Finfo(eps=0.00097656, tiny=6.1035e-05),
+    'torch.FloatTensor': _Finfo(eps=1.19209e-07, tiny=1.17549e-38),
+    'torch.DoubleTensor': _Finfo(eps=2.22044604925e-16, tiny=2.22507385851e-308),
+    'torch.cuda.HalfTensor': _Finfo(eps=0.00097656, tiny=6.1035e-05),
+    'torch.cuda.FloatTensor': _Finfo(eps=1.19209e-07, tiny=1.17549e-38),
+    'torch.cuda.DoubleTensor': _Finfo(eps=2.22044604925e-16, tiny=2.22507385851e-308),
+}
+
+
+def _finfo(tensor):
+    """
+    Return floating point info about a `Tensor` or `Variable`:
+    - `.eps` is the smallest number that can be added to 1 without being lost.
+    - `.tiny` is the smallest positive number greater than zero
+      (much smaller than `.eps`).
+    """
+    return _FINFO[tensor.type()]
```

Comment:
let's use `type(tensor)` to avoid building up the string description and matching on that (`type(tensor)` will simply hash and compare the pointer value)

Reply:
Won't that fail for `Variable`s and subclasses like `Parameter`s? Somehow we need to traverse a class hierarchy, e.g. in `.type` method resolution.

Reply:
I'll just memoize by `type()` and fall back to `.type()`

Reply:
Well `type()` just won't work since all `Variable`s have the same type, so I've used `.storage_type()` instead.